README
NAME
    Pod::Weaver::PluginBundle::ReplaceBoilerplate - A Pod::Weaver bundle for
    replacing the boilerplate in a Pod document.

VERSION
    version 0.99_02

OVERVIEW
    This is a plugin bundle intended to replace use of [@Default] with
    equivilent behaviour but with Pod::Weaver::Role::SectionReplacer enabled
    plugins.

    It is nearly equivalent to the following:

      [@CorePrep]

      [ReplaceName]
      [ReplaceVersion]

      [Leftovers]

      [ReplaceAuthors]
      [ReplaceLegal]

INSTALLATION
    To install this module, run the following commands:

      perl Build.PL
      ./Build
      ./Build test
      ./Build install

SUPPORT AND DOCUMENTATION
    After installing, you can find documentation for this module with the
    perldoc command.

      perldoc Pod::Weaver::PluginBundle::ReplaceBoilerplate

    You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Pod-Weaver-PluginBundle-Rep
        laceBoilerplate

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Pod-Weaver-PluginBundle-ReplaceBoilerplate

    CPAN Ratings
        http://cpanratings.perl.org/d/Pod-Weaver-PluginBundle-ReplaceBoilerp
        late

    Search CPAN
        http://search.cpan.org/dist/Pod-Weaver-PluginBundle-ReplaceBoilerpla
        te/

AUTHOR
    Sam Graham <libpod-weaver-pluginbundle-replaceboilerplate-perl BLAHBLAH
    illusori.co.uk>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2010 by Sam Graham
    <libpod-weaver-pluginbundle-replaceboilerplate-perl BLAHBLAH
    illusori.co.uk>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Pod-Weaver-PluginBundle-ReplaceBoilerplate

0.99_02 Version released at 2010-09-13-14:05.  *** Development version. ***
        Require Moose::Autobox 0.11 for $array->first().
        Require Pod::Weaver 3.101630 to get expected Authors plugin behavior.
        Require Pod::Elemental 0.101620.
        Made plugin abstracts consistent.
        Expanded overview documentation to include links to base plugins.

0.99_01 Version released at 2010-09-12-16:10.  *** Development version. ***
        Initial release.


lib/Pod/Weaver/Section/ReplaceName.pm
package Pod::Weaver::Section::ReplaceName;

# ABSTRACT: Add or replace a NAME section with abstract.

use Moose;

extends 'Pod::Weaver::Section::Name';
with 'Pod::Weaver::Role::SectionReplacer';

our $VERSION = '0.99_02';

sub default_section_name { 'NAME' }

1;

__END__


=pod

=head1 NAME

Pod::Weaver::Section::ReplaceName - Add or replace a NAME section with abstract.

=head1 VERSION

version 0.99_02

=head1 OVERVIEW

This section plugin provides the same behaviour as
L<Pod::Weaver::Section::Name> but with the
L<Pod::Weaver::Role::SectionReplacer> role applied.

This section plugin will produce a hunk of Pod giving the name of the document
as well as an abstract, like this:

  =head1 NAME

  Some::Document - a document for some

It will determine the name and abstract by inspecting the C<ppi_document>
input parameter.
It will look for the first package declaration, and for a comment in this form:

  # ABSTRACT: a document for some

=for readme stop

=begin :internal

=head1 INTERNAL METHODS

=over

=item default_section_name

Gives the name used as the heading for this section.

=back

=end :internal

=for readme continue

=head1 AUTHOR

Sam Graham <libpod-weaver-pluginbundle-replaceboilerplate-perl BLAHBLAH illusori.co.uk>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Sam Graham <libpod-weaver-pluginbundle-replaceboilerplate-perl BLAHBLAH illusori.co.uk>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Pod/Weaver/Section/ReplaceLegal.pm
package Pod::Weaver::Section::ReplaceLegal;

# ABSTRACT: Add or replace a COPYRIGHT AND LICENSE section.

use Moose;

extends 'Pod::Weaver::Section::Legal';
with 'Pod::Weaver::Role::SectionReplacer';

our $VERSION = '0.99_02';

has year => (
  is  => 'ro',
  isa => 'Str',
  );

around weave_section => sub
{
    my ( $orig, $self, $document, $input ) = @_;

    if( my $year = $self->year )
    {
        return unless $input->{license};
        $year =~ s/current/(localtime)[ 5 ] + 1900/e;
        $input->{ license }->{ year } = $year;
    }

    return( $self->$orig( $document, $input ) );
};

sub default_section_name { 'COPYRIGHT AND LICENSE' }
sub default_section_aliases
{
    [
        'LICENSE AND COPYRIGHT',
        'LICENSE & COPYRIGHT',
        'COPYRIGHT & LICENSE',
    ]
}

no Moose;
1;

__END__


=pod

=head1 NAME

Pod::Weaver::Section::ReplaceLegal - Add or replace a COPYRIGHT AND LICENSE section.

=head1 VERSION

version 0.99_02

=head1 OVERVIEW

This section plugin provides the same behaviour as
L<Pod::Weaver::Section::Legal> but with the
L<Pod::Weaver::Role::SectionReplacer> role applied.

It will produce a hunk of Pod giving the copyright and license
information for the document, like this:

  =head1 COPYRIGHT AND LICENSE

  This document is copyright (C) 1991, Ricardo Signes.

  This document is available under the blah blah blah.

This plugin will do nothing if no C<license> input parameter is available.  The
C<license> is expected to be a L<Software::License> object.

=head1 CUSTOMIZATION

In addition to the standard L<Pod::Weaver::Section::Legal> behaviour,
you may customize the behaviour of L<Pod::Weaver::Section::ReplaceLegal>
with the following options in your C<weaver.ini>:

=over

=item B<year> = I<year string>

You may supply a copyright year, to be used by L<Software::License> via the
C<year> parameter.

The year may be a number or it may be an arbitrary string.

Within any string the word 'current' will be replaced by the current year.

For example in your C<weaver.ini>:

  [ReplaceLegal]
  year = 2005-current

This will, in the year 2010, produce the year string 2005-2010 for use
within the L<Software::License> object.

=back

=for readme stop

=begin :internal

=head1 INTERNAL METHODS

=over

=item default_section_name

Gives the name used as the heading for this section.

=item default_section_aliases

Gives alternative names that an existing section might be using.

=back

=end :internal

=for readme continue

=head1 AUTHOR

Sam Graham <libpod-weaver-pluginbundle-replaceboilerplate-perl BLAHBLAH illusori.co.uk>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Sam Graham <libpod-weaver-pluginbundle-replaceboilerplate-perl BLAHBLAH illusori.co.uk>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Pod/Weaver/Section/ReplaceAuthors.pm
package Pod::Weaver::Section::ReplaceAuthors;

# ABSTRACT: Add or replace an AUTHOR or AUTHORS section.

use Moose;

extends 'Pod::Weaver::Section::Authors';
with 'Pod::Weaver::Role::SectionReplacer';

our $VERSION = '0.99_02';

sub default_section_name { 'AUTHORS' }
sub default_section_aliases { [ 'AUTHOR' ] }

no Moose;
1;

__END__

=pod

=head1 NAME

Pod::Weaver::Section::ReplaceAuthors - Add or replace an AUTHOR or AUTHORS section.

=head1 VERSION

version 0.99_02

=head1 OVERVIEW

This section plugin provides the same behaviour as
L<Pod::Weaver::Section::Authors> but with the
L<Pod::Weaver::Role::SectionReplacer> role applied.

It will add or replace a listing of the document's authors.
It expects the C<authors> input parameter to be an arrayref of strings.
If no C<authors> parameter is given, it will do nothing.
Otherwise, it produces a hunk like this:

  =head1 AUTHOR

  Author <a@example.com>

Or in the case of multiple authors:

  =head1 AUTHORS

  =over
  
  =item Author One <a1@example.com>

  =item Author Two <a2@example.com>

  =back

=for readme stop

=begin :internal

=head1 INTERNAL METHODS

=over

=item default_section_name

Gives the name used as the heading for this section.

=item default_section_aliases

Gives alternative names that an existing section might be using.

=back

=end :internal

=for readme continue

=head1 AUTHOR

Sam Graham <libpod-weaver-pluginbundle-replaceboilerplate-perl BLAHBLAH illusori.co.uk>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Sam Graham <libpod-weaver-pluginbundle-replaceboilerplate-perl BLAHBLAH illusori.co.uk>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Pod/Weaver/Section/ReplaceVersion.pm
package Pod::Weaver::Section::ReplaceVersion;

use Moose;

# ABSTRACT: Add or replace a VERSION section.

extends 'Pod::Weaver::Section::Version';
with 'Pod::Weaver::Role::SectionReplacer';

our $VERSION = '0.99_02';

sub default_section_name { 'VERSION' }

no Moose;
1;

__END__


=pod

=head1 NAME

Pod::Weaver::Section::ReplaceVersion - Add or replace a VERSION section.

=head1 VERSION

version 0.99_02

=head1 OVERVIEW

This section plugin provides the same behaviour as
L<Pod::Weaver::Section::Version> but with the
L<Pod::Weaver::Role::SectionReplacer> role applied.

It will produce a hunk of Pod meant to indicate the version of
the document being viewed, like this:

  =head1 VERSION

  version 1.234

It will do nothing if there is no C<version> entry in the input.

=for readme stop

=begin :internal

=head1 INTERNAL METHODS

=over

=item default_section_name

Gives the name used as the heading for this section.

=back

=end :internal

=for readme continue

=head1 AUTHOR

Sam Graham <libpod-weaver-pluginbundle-replaceboilerplate-perl BLAHBLAH illusori.co.uk>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Sam Graham <libpod-weaver-pluginbundle-replaceboilerplate-perl BLAHBLAH illusori.co.uk>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Pod/Weaver/PluginBundle/ReplaceBoilerplate.pm
package Pod::Weaver::PluginBundle::ReplaceBoilerplate;

# ABSTRACT: A Pod::Weaver bundle for replacing the boilerplate in a Pod document.

use strict;
use warnings;

use namespace::autoclean;

our $VERSION = '0.99_02';

use Pod::Weaver::Config::Assembler;
sub _exp { Pod::Weaver::Config::Assembler->expand_package($_[0]) }

sub mvp_bundle_config {
  return (
    [ '@ReplaceBoilerplate/CorePrep',         _exp('@CorePrep'),        {} ],
    [ '@ReplaceBoilerplate/ReplaceName',      _exp('ReplaceName'),      {} ],
    [ '@ReplaceBoilerplate/ReplaceVersion',   _exp('ReplaceVersion'),   {} ],

    [ '@ReplaceBoilerplate/Leftovers',        _exp('Leftovers'),        {} ],

    [ '@ReplaceBoilerplate/ReplaceAuthors',   _exp('ReplaceAuthors'),   {} ],
    [ '@ReplaceBoilerplate/ReplaceLegal',     _exp('ReplaceLegal'),     {} ],
  )
}

1;

__END__

=pod

=head1 NAME

Pod::Weaver::PluginBundle::ReplaceBoilerplate - A Pod::Weaver bundle for replacing the boilerplate in a Pod document.

=head1 VERSION

version 0.99_02

=head1 OVERVIEW

This is a plugin bundle intended to replace use of [@Default] with equivilent
behaviour but with L<Pod::Weaver::Role::SectionReplacer> enabled plugins.

It is nearly equivalent to the following:

  [@CorePrep]

  [ReplaceName]
  [ReplaceVersion]

  [Leftovers]

  [ReplaceAuthors]
  [ReplaceLegal]

=begin readme

=head1 INSTALLATION

To install this module, run the following commands:

  perl Build.PL
  ./Build
  ./Build test
  ./Build install

=end readme

=for readme stop

=begin :internal

=head1 INTERNAL METHODS

=over

=item mvp_bundle_config

Returns the config data structure to substitute for this PluginBundle.

=back

=end :internal

=for readme continue

=head1 SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

  perldoc Pod::Weaver::PluginBundle::ReplaceBoilerplate

You can also look for information at:

=over

=item RT, CPAN's request tracker

http://rt.cpan.org/NoAuth/Bugs.html?Dist=Pod-Weaver-PluginBundle-ReplaceBoilerplate

=item AnnoCPAN, Annotated CPAN documentation

http://annocpan.org/dist/Pod-Weaver-PluginBundle-ReplaceBoilerplate

=item CPAN Ratings

http://cpanratings.perl.org/d/Pod-Weaver-PluginBundle-ReplaceBoilerplate

=item Search CPAN

http://search.cpan.org/dist/Pod-Weaver-PluginBundle-ReplaceBoilerplate/

=back

=head1 AUTHOR

Sam Graham <libpod-weaver-pluginbundle-replaceboilerplate-perl BLAHBLAH illusori.co.uk>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Sam Graham <libpod-weaver-pluginbundle-replaceboilerplate-perl BLAHBLAH illusori.co.uk>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'Pod::Weaver::PluginBundle::ReplaceBoilerplate',
    license             => 'perl',
    dist_author         => q{Sam Graham <libpod-weaver-pluginbundle-replaceboilerplate-perl@illusori.co.uk>},
    dist_version_from   => 'lib/Pod/Weaver/PluginBundle/ReplaceBoilerplate.pm',
    build_requires => {
        'Test::More' => 0,
    },
    add_to_cleanup      => [ 'Pod-Weaver-PluginBundle-ReplaceBoilerplate-*' ],
    requires       => {
        'Moose'          => 0,
        'Moose::Autobox' => '0.11',
        'Pod::Elemental' => 0,
        'Pod::Weaver'    => '3.101630',
        'Pod::Weaver::Role::SectionReplacer' => 0,
        'Software::License' => 0,
    },
    meta_merge => {
        'resources' => {
            'repository' => 'http://github.com/illusori/Perl-Pod-Weaver-PluginBundle-ReplaceBoilerplate',
            },
        },
    create_readme => 1,
    sign => 1,
    dynamic_config => 0,
);

$builder->create_build_script();
